### ISSUE TITLE
Is there a mode that does not rely on host mpi?

### ISSUE THREAD

**Issue by chuqing1997** [apptainer/apptainer#1413]
### Version of Apptainer

# apptainer --version
apptainer version 1.1.8

### Expected behavior

Is there any other way to run MPI programs using the Apptainer container, apart from the Bind model and Hybrid model? For example, mpi is only provided by containers, but not in the host. Because I want to run the Apptainer container without relying on the host's mpi environment.

Is there a mode that does not rely on host mpi?

Sincerely thank you for your answer.

**Comment by GodloveD** [comment]
If you have a tool on your host system that can use the PMI standard to accomplish the wire-up process than yes, you can use that instead.  Something like Slurm will do the trick if it was built with PMI support.  You also need to make sure that your containerized MPI was built with PMI support.  I'm actually in the process of writing a blogpost about this exact topic right now!  In the meantime there are a few talks that will guide you through the process:
https://youtu.be/VkNRQ8YsdE8?t=3554
https://youtu.be/FvyDEnpa9vM

**Comment by chuqing1997** [comment]
> 

Thank you again, boss, for your reply. I watched Jonathan Anderson's video and gained some understanding, but due to my English and personal proficiency, I did not fully grasp it. I would like to ask you a few more questions. I hope you can answer in your busy schedule.

1.What's mean "Slurm built with PMI support"?Is need a specific version of Slurm? Or need to add plugins to Slurm?Or does it mean that my srun command supports as long as it has the option "-- mpi="?

2."The MPI in the container is built using PMI support", I understand the meaning of this sentence. However, in the video, the sif in the demo is built by writing the program we need in the" spack. yaml "file and then using "spack containerize". However, due to my personal level, I do not quite understand the principle of spack, or I am not accustomed to using spack, so I would like to know how to create a sif that supports PMI through def files, May I request you to send a def file version of the "Hello World" program?

3.Where will your blog be posted after writing? Where can I see it? Looking forward to it very much.

Because of my poor English proficiency, I don't know if the words used on it are stiff or not,but I am seeking your advice with a grateful, respectful, sincere, apprentice, and hopeful attitude. I hope you can teach me! Thank you!

**Comment by chuqing1997** [comment]
> If you have a tool on your host system that can use the PMI standard to accomplish the wire-up process than yes, you can use that instead. Something like Slurm will do the trick if it was built with PMI support. You also need to make sure that your containerized MPI was built with PMI support. I'm actually in the process of writing a blogpost about this exact topic right now! In the meantime there are a few talks that will guide you through the process: https://youtu.be/VkNRQ8YsdE8?t=3554 https://youtu.be/FvyDEnpa9vM

I have completed this process through your guidance and video. Thank you. But there is a new issue that arises if I want to run both a multi threaded MPI implementation and a single threaded program that is not an MPI implementation within an apptainer container (such as running echo "Thank you!" only once).

The command is as follows:

cat/opt/test.sh

/opt/mpitest
echo "Thank you

srun -- mpi=pmi2- N 2-- ntasks=4-- ntasks per node=2 apptainer exec test.sif/opt/test.sh

This command will execute the/opt/mpitest and echo commands 4 times, outputting 4 lines of 'Thank you!'. And I just want to print a 'Thank you!', what should I do?

Is it impossible to achieve this requirement within an apptainer container?

Thank you for your answer

**Comment by GodloveD** [comment]
Sorry for the delay.  The blog post is here.

https://ciq.com/blog/a-new-approach-to-mpi-in-apptainer/

**Comment by GodloveD** [comment]
As far as your second question goes, I assume you have probably figured something out already, but just in case, it looks like something you will need to split out into a different command (and I different call to Apptainer).
